---
title: Initital processing of SARS-COV2 Infection samples for Trex Manuscript
author: 
- name: Nick Borcherding
  email: ncborch@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
date: "August 19, 2020"
output:
  BiocStyle::html_document:
    toc_float: true

---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Introduction


##Computer Info:
Analysis is being run on a 13-inch Macbook Pro (2017) with 3.5 GHz Dual-Core i7 and 16 GB of 2133 MHz LPDDR3. There is a limit to the speed and RAM, so for the purposes of the markdown file, I will not run some of the code blocks during the kniting (look at eval = F). You will also see the call *rm()* frequently to just remove non-vital objects from the global environment.

***

# Loading Libraries

In general I like to load libraries here that we will use universally, and then call other libraries when we need them in the code chunks that are relevant. 

```{r}
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(scDblFinder))
suppressPackageStartupMessages(library(BiocParallel))
suppressPackageStartupMessages(library(Trex))

```

I also like to set a color palette before I begin - this way all the colors are consistent throughout the publication figures.

```{r setup, include=FALSE}
library(viridis)
colorblind_vector <- viridis_pal()(7)
```


***

# Loading and Processing the Data


## Load Data
```{r eval=FALSE}
dir.create("dataExplore")
dir.create("dataExplore/qc")
"%!in%" <- Negate("%in%")
file_list <- list.files("./data/SequencingRuns", pattern = "Sar")
list <- NULL
for (i in seq_along(file_list)){
    tmp <-  Read10X(paste0("./data/SequencingRuns/", file_list[i]))
    s.obj <- CreateSeuratObject(tmp$`Gene Expression`, project = file_list[i])
    row.index <- which(rownames(tmp$`Antibody Capture`) %!in% c("TCRVa7.2_TotalSeqC","TCRVa24Ja18_TotalSeqC"))
    
    adt_assay <- CreateAssayObject(counts = tmp$`Antibody Capture`[row.index,])
    
  
    # add this assay to the previously created Seurat object
    s.obj[["ADT"]] <- adt_assay
    s.obj <- RenameCells(object = s.obj, new.names = paste0(file_list[i], "_", rownames(s.obj[[]])))
    s.obj<- subset(s.obj, subset = nFeature_RNA > 100)
    s.obj[["mito.genes"]] <- PercentageFeatureSet(s.obj, pattern = "^MT-")
    
    p1 <- VlnPlot(object = s.obj, features = c("nCount_RNA")) + theme(legend.position = "none")
    p2 <- VlnPlot(object = s.obj, features = c("nFeature_RNA")) + theme(legend.position = "none")
    p3 <- VlnPlot(object = s.obj, features = c("mito.genes")) + theme(legend.position = "none")
    
    pdf(paste0("./dataExplore/qc/", file_list[i], ".pdf"), height = 8, width=12)
    grid.arrange(p1, p2, p3, ncol = 3)
    dev.off()
    
    ###########################
  #Here is the filtering step
  ############################
  standev <- sd(log(s.obj$nFeature_RNA))*2.5 #cutting off above standard deviation of 2.5
  mean <- mean(log(s.obj$nFeature_RNA))
  cut <- round(exp(standev+mean))
  s.obj <- subset(s.obj, subset = mito.genes < 10 & nFeature_RNA < cut)
    
   ###########################################
  #Estimate Doublets for Each Sequencing Run
  ############################################
  sce <- as.SingleCellExperiment(s.obj)
  sce <- scDblFinder(sce, BPPARAM=MulticoreParam(3))
  doublets <- data.frame(db.weight.score = sce$scDblFinder.score, db.ratio = sce$scDblFinder.weighted, 
                         db.class = sce$scDblFinder.class, db.score = sce$scDblFinder.score)
  rownames(doublets) <- rownames(sce@colData)
  s.obj <- AddMetaData(s.obj, doublets)
  rm(sce)
  
  ####Adding meta data
  directory <- readxl::read_xlsx("./data/sample.directory.xlsx") #Meta.data
  meta <- s.obj[[]]
  rownames <- rownames(meta)
  meta <- merge(meta, directory, by.x = "orig.ident", by.y = "ID")
  meta <- meta[,9:ncol(meta)]
  rownames(meta) <- rownames
  
  
  ## Finding variable features and normalizing data for integration to follow
  s.obj <- AddMetaData(s.obj, meta)
  s.obj <- NormalizeData(s.obj)
  s.obj <- FindVariableFeatures(s.obj, selection.method = "vst", nfeatures = 2000)
  s.obj <- quietTCRgenes(s.obj)
  VariableFeatures(s.obj, assay = "ADT") <- rownames(s.obj[["ADT"]])
  s.obj <- NormalizeData(s.obj, normalization.method = 'CLR', margin = 2, assay = "ADT")

  list[[i]] <- s.obj
}
saveRDS(list, file = "./data/filtered_seuratObjects.rds")
rm(tmp)
rm(adt_assay)
rm(s.obj)
```


## Integrate Assays

```{r}

#####################
#Integrating RNA
###################

# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = list)

immune.anchors <- FindIntegrationAnchors(object.list = list, anchor.features = features)
# this command creates an 'integrated' data assay
immune.combined <- IntegrateData(anchorset = immune.anchors)
immune.combined <- ScaleData(immune.combined, verbose = FALSE)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)

saveRDS(immune.combined, file = "./data/RNAintegrated_seuratObjects.rds")
rm(features)
rm(immune.anchors)
rm(immune.combined)


############################
#Integrating ADT
############################
list <- readRDS("./data/filtered_seuratObjects.rds")
features <- rownames(list[[1]][["ADT"]])

list <- lapply(X = list, FUN = function(x) {
    DefaultAssay(x) <- "ADT"
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE)
})

ADT.anchors <- FindIntegrationAnchors(object.list = list, reference = c(1, 2), reduction = "rpca", 
    dims = 1:30, assay = rep("ADT", 14), anchor.features = 37)
ADT.integrated <- IntegrateData(anchorset = ADT.anchors, dims = 1:30)
ADT <- ADT.integrated@assays$integrated
rm(ADT.anchors)
rm(ADT.integrated)

############################
#Combining all to single object
############################
list <- readRDS("./data/RNAintegrated_seuratObjects.rds")
list[["ADT.integrated"]] <- ADT
saveRDS(list, file = "./data/Fullintegrated_seuratObjects.rds")
```

```{r}
library(scRepertoire)
file_list <- list.files("./data/SequencingRuns", pattern = "Sar")
######################################
#iterate to make a list of contig csvs
######################################


contig.list <- NULL
for (i in seq_along(file_list)){
  contig.list[[i]] <-  read.csv(paste0("./data/SequencingRuns/", file_list[i], "/contig_annotations.csv"))
}
names(contig.list) <- file_list

##################################################
#Reducing the data to the individual barcode level
##################################################
combinedObject <- combineTCR(contig.list, samples = file_list, filterMulti = TRUE, removeNA = TRUE, cells = "T-AB")
combinedObject <- addVariable(combinedObject, name = "Patient", variables = substr(file_list, 1,5))
list <- combineExpression(combinedObject, list, cloneCall = "aa", groupBy = "Patient")
saveRDS(list, file = "./data/Filteredintegrated_seuratObjects.rds")
```




